### PR TITLE
Artifact light interaction changes

### PR DESCRIPTION
**Pull request by i2amroy**
Changes the way artifacts interact with natural light. This might cause a slight performance hit when below ground, but allows for events to change the light level regardless of the player's z-level. Fixes the "flash" artifact to actually work belowground.

### CODE REVIEW

**Comment by i2amroy** [comment]
It seems that lua is freaking out here and breaking the build on this despite this seeming like a pretty straightforward change. At first glance it seems like the build is still trying to do something with a function that I removed, and I can honestly say I have no idea why.

**Comment by narc0tiq** [comment]
It seems that `lua/class_definitions.lua` still references that member you removed. I'm putting together a PR to fix that for you.

File: src/game.cpp [hunk 1]
```diff
@@ -5389,48 +5389,60 @@ void game::hallucinate( const tripoint &center )
...
 
-float game::ground_natural_light_level() const
+float game::natural_light_level() const
 {
     // Already found the light level for now?
     if( calendar::turn == latest_lightlevel_turn ) {
         return latest_lightlevel;
     }
+    // Should be completely overwritten next, thus no initialization value here.
+    float ret;
```
Comment (kevingranade) [inline src/game.cpp]: This assumption seems to be invalidated underground, it's generally better to initialize to something sensible no matter what's happening later.

Comment (i2amroy) [inline src/game.cpp]: Fair enough.

**Comment by Coolthulhu** [comment]
Not directly related to this PR, but:

Damn, our artifact interface sure is outdated. And buggy too - my game somehow generated an artifact like this:
```
{"type":"artifact_tool","id":"bot_copbot","name":"undefined-bot_copbots","description":"A strange shimmering... nothing.  You think it wants to be a bot_copbot.","sym":46,"color":139264,"price":0,"materials":[],"volume":0,"weight":0,"melee_dam":0,"melee_cut":0,"m_to_hit":0,"item_flags":[],"techniques":[],"ammo":"NULL","max_charges":0,"def_charges":0,"charges_per_use":1,"turns_per_charge":0,"revert_to":"null","charge_type":0,"effects_wielded":[],"effects_activated":[],"effects_carried":[]}
```
It is related to my merge branch (other PR I tested had bad jsons). This seems to have bugged my test world and cause it to crash on load.

Passive artifact enum is enums.h, while active one is in artifact.h. The whole idea of artifacts being saved to a separate file rather than a property of item makes it all non-standard and hacky. The way artifact properties are saved (as enums cast to `int`s) makes it way harder to find what they do.

Not directly a problem with this PR, just makes it harder to test.

**Comment by Coolthulhu** [comment]
Due to the recent save bug, I have no idea how would I test this PR, as I don't know of any in-game artifact editor. Deassigning for now.

**Comment by BevapDin** [comment]
> Passive artifact enum is enums.h, while active one is in artifact.h. 

Passive ones are accessed all over the place, you can find `u.has_artifact_with(XXX)` everywhere. Therefor the enum needs to be exposed. Putting it in artifact.h would require include that header, which in turn included itype.h and a bunch of other headers, which was not nice. There for I moved that one enum to enums.h

Active don't need this, they are only uses in a few places (game.cpp, iuse.cpp)

----

> my game somehow generated an artifact like this:

This item is created by the `item_factory` when an undefined itype is requested. See the name "undefined-bot_copbots", somehow an item id `bot_copbots` has been requested.

The class creates a new artifact type because the artifact classes/system already supports storing an item type definition in the saves.

**Pull request merged**